src/game/logic/object/weaponset.cpp

```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
+                        is_in_range = weapon->Is_Source_Object_With_Goal_Position_Within_Attack_Range(
+                            source_obj, &garrison_pos, victim_obj, &loc);
+                        is_garrison = true;
+                    }
+                } else {
+                    is_in_range = victim_obj != nullptr ? weapon->Is_Within_Attack_Range(source_obj, victim_obj) :
+                                                          weapon->Is_Within_Attack_Range(source_obj, location);
+                }
+
+                if (is_in_range) {
+                    weapon_in_range = true;
+                    break;
+                }
+            }
+        }
+    }
+
+    if ((source_obj->Is_KindOf(KINDOF_IMMOBILE) || source_obj->Is_KindOf(KINDOF_SPAWNS_ARE_THE_WEAPONS)
+            || containedby != nullptr)
+        && has_weapon && !weapon_in_range && type != ATTACK_TYPE_4) {
+        return ATTACK_RESULT_1;
+    }
+
+    CanAttackResult result = static_cast<CanAttackResult>(is_in_range + 2);
+
+    if (Has_Any_Damage_Weapon()) {
+        if ((mask & m_totalAntiMask) == 0) {
+            return ATTACK_RESULT_1;
+        }
+
+        if (victim_obj == nullptr) {
+            return result;
+        }
+
+        if (!Is_Any_Within_Target_Pitch(source_obj, victim_obj)) {
+            return ATTACK_RESULT_1;
+        }
+
+        WeaponSlotType start_slot;
+        WeaponSlotType end_slot;
+
+        if (Is_Cur_Weapon_Locked()) {
+            start_slot = m_curWeapon;
+            end_slot = m_curWeapon;
+        } else if (wslot == WEAPONSLOT_UNK) {
+            start_slot = WEAPONSLOT_TERTIARY;
+            end_slot = WEAPONSLOT_PRIMARY;
+        } else {
+            start_slot = wslot;
+            end_slot = wslot;
+        }
+
+        for (int i = start_slot; i >= end_slot; i--) {
+            Weapon *weapon2 = m_weapons[i];
+
+            if (weapon2 != nullptr && weapon2->Estimate_Weapon_Damage(source_obj, victim_obj) != 0.0f
+                && (weapon2->Get_Damage_Type() != DAMAGE_KILL_PILOT || !source_obj->Is_KindOf(KINDOF_HERO)
+                    || m_curWeapon != WEAPONSLOT_PRIMARY || wslot != WEAPONSLOT_UNK)) {
+                return result;
+            }
+        }
+    }
+
+    ContainModuleInterface *contain = source_obj->Get_Contain();
+
+    if (contain != nullptr) {
+        if (contain->Is_Passenger_Allowed_To_Fire(OBJECT_UNK)) {
+            const std::list<Object *> *list = contain->Get_Contained_Items_List();
+
+            if (list != nullptr) {
+                for (auto it = list->begin(); it != list->end(); it++) {
+                    Object *obj = *it;
+
+                    if (obj->Is_Able_To_Attack()) {
+                        CanAttackResult can =
+                            obj->Get_Able_To_Use_Weapon_Against_Target(type, victim_obj, location, source, WEAPONSLOT_UNK);
+
+                        if (can == ATTACK_RESULT_3 || can == ATTACK_RESULT_2) {
```

Comment:
Perhaps swap to make it easier on the brain
```
can == ATTACK_RESULT_2 || can == ATTACK_RESULT_3
```

Reply:
fixed.